core-web/libs/block-editor/src/lib/extensions/dot-placeholder/dot-placeholder-plugin.ts

```diff
@@ -0,0 +1,114 @@
+                props: {
+                    decorations: ({ doc, selection }) => {
+                        const active = this.editor.isEditable || !this.options.showOnlyWhenEditable;
+
+                        const { anchor } = selection;
+                        const decorations: Decoration[] = [];
+
+                        if (!active) {
+                            return null;
+                        }
+
+                        // only calculate isEmpty once due to its performance impacts (see issue #3360)
+                        const emptyDocInstance = doc.type.createAndFill();
+                        const isEditorEmpty =
+                            emptyDocInstance?.sameMarkup(doc) &&
+                            emptyDocInstance.content.findDiffStart(doc.content) === null;
+
+                        doc.descendants((node, pos) => {
+                            const hasAnchor = anchor >= pos && anchor <= pos + node.nodeSize;
+                            const isEmpty = !node.isLeaf && !node.childCount;
+
+                            if ((hasAnchor || !this.options.showOnlyCurrent) && isEmpty) {
+                                const classes = [this.options.emptyNodeClass];
+
+                                if (isEditorEmpty) {
+                                    classes.push(this.options.emptyEditorClass);
+                                }
+
+                                const decoration = Decoration.widget(pos, () => {
+                                    const div = document.createElement('div');
+
+                                    if (node.type.name === NodeTypes.HEADING) {
+                                        const title = document.createElement(
+                                            `h${node.attrs.level}`
+                                        );
+                                        title.dataset.placeholder = `${toTitleCase(
+                                            node.type.name
+                                        )} ${node.attrs.level}`;
+                                        title.classList.add('is-empty');
+                                        div.appendChild(title);
+                                    } else {
+                                        const paragraph = document.createElement('p');
+                                        paragraph.dataset.placeholder = this.options.placeholder;
+                                        paragraph.classList.add(...classes);
+                                        div.appendChild(paragraph);
+                                    }
+
+                                    const button = document.createElement('button');
+                                    button.classList.add('add-button');
+                                    button.style.position = 'absolute';
+                                    button.style.left = '20px';
+                                    button.innerHTML = '<i class="material-icons">add</i>';
```

Comment:
Can we create a `function` for this? A `function` that returns the `button` to make it more readable.